core/src/main/scala/com/microsoft/azure/synapse/ml/explainers/ICEExplainer.scala

```diff
@@ -0,0 +1,252 @@
+  )
+
+  def getKind: String = $(kind)
+  def setKind(value: String): this.type = set(kind, value)
+
+  setDefault(kind -> "individual", numericFeatures -> Seq.empty[ICENumericFeature],
+    categoricalFeatures -> Seq.empty[ICECategoricalFeature])
+}
+
+/**
+  * ICETransformer displays the model dependence on specified features with the given dataframe
+  * as background dataset. It supports 2 types of plots: individual - dependence per instance and
+  * average - across all the samples in the dataset.
+  * Note: This transformer only supports one-way dependence plot.
+  */
+@org.apache.spark.annotation.Experimental
+class ICETransformer(override val uid: String) extends Transformer
+  with HasExplainTarget
+  with HasModel
+  with ICEFeatureParams
+  with Wrappable
+  with ComplexParamsWritable {
+
+  override protected lazy val pyInternalWrapper = true
+
+  def this() = {
+    this(Identifiable.randomUID("ICETransformer"))
+  }
+
+  private def calcDependence(df: DataFrame, idCol: String, targetClassesColumn: String,
+                             feature: String, values: Array[_], outputColName: String): DataFrame = {
+
+    val dataType = df.schema(feature).dataType
+    val explodeFunc = explode(array(values.map(v => lit(v)): _*).cast(ArrayType(dataType)))
+
+    val predicted = getModel.transform(df.withColumn(feature, explodeFunc))
+    val targetCol = DatasetExtensions.findUnusedColumnName("target", predicted)
+    val dependenceCol = DatasetExtensions.findUnusedColumnName("feature__dependence", predicted)
+
+    val explainTarget = extractTarget(predicted.schema, targetClassesColumn)
+    val result = predicted.withColumn(targetCol, explainTarget)
+
+    getKind.toLowerCase match {
+      case this.averageKind =>
+        // PDP output schema: 1 row * 1 col (pdp for the given feature: feature_value -> explanations)
+        result
+          .groupBy(feature)
+          .agg(Summarizer.mean(col(targetCol)).alias(dependenceCol))
+          .agg(
+            map_from_arrays(
+              collect_list(feature),
+              collect_list(dependenceCol)
+            ).alias(outputColName)
+          )
+
+      case this.individualKind =>
+        // ICE output schema: n rows * 2 cols (idCol + ice for the given feature: map(feature_value -> explanations))
+        result
+          .groupBy(idCol)
+          .agg(
+            map_from_arrays(
+              collect_list(feature),
+              collect_list(targetCol)
+            ).alias(outputColName)
+          )
+    }
+  }
+
+  def transform(ds: Dataset[_]): DataFrame = {
+    transformSchema(ds.schema)
+    val df = ds.toDF
+    val idCol = DatasetExtensions.findUnusedColumnName("idCol", df)
+    val targetClasses = DatasetExtensions.findUnusedColumnName("targetClasses", df)
+    val dfWithId = df
+      .withColumn(idCol, monotonically_increasing_id())
+      .withColumn(targetClasses, this.get(targetClassesCol).map(col).getOrElse(lit(getTargetClasses)))
```

Comment:
nit: use getters directly here

Reply:
targetClassesCol is Optional by design, that's why we're using get like this